### PR TITLE
Resolve profiles

### PR DESCRIPTION
**Pull request by trapacska**

### CODE REVIEW

File: bitrise.yml [hunk 1]
```diff
@@ -36,6 +36,7 @@ workflows:
     - SAMPLE_APP_URL: https://github.com/bitrise-io/sample-apps-ios-simple-objc.git
     - BITRISE_PROJECT_PATH: ios-simple-objc/ios-simple-objc.xcodeproj
     - BITRISE_SCHEME: ios-simple-objc
+    - FORCE_CODE_SIGN_IDENTITY: "iPhone Developer: Bitrise Bot"
```
Comment (godrei) [inline bitrise.yml]: remove this

File: bitrise.yml [hunk 2]
```diff
@@ -46,6 +47,7 @@ workflows:
     - BITRISE_PROJECT_PATH: ios-xcode-8.0/ios-xcode-8.0.xcodeproj
     - BITRISE_SCHEME: ios-xcode-8.0
     - FORCE_TEAM_ID: 72SA8V3WYL
+    - FORCE_CODE_SIGN_IDENTITY: "iPhone Developer: Bitrise Bot"
```
Comment (godrei) [inline bitrise.yml]: remove this

File: bitrise.yml [hunk 3]
```diff
@@ -89,6 +91,7 @@ workflows:
     - path::./:
         inputs:
         - output_tool: xcodebuild
+        - team_id: 72SA8V3WYL
```
Comment (godrei) [inline bitrise.yml]: remove this

File: utils/code_sign_mapping.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	glob "github.com/ryanuber/go-glob"
+
+	"github.com/bitrise-io/steps-certificate-and-profile-installer/certificateutil"
+	"github.com/bitrise-io/steps-certificate-and-profile-installer/profileutil"
+)
+
+// ProfileGroup ...
+type ProfileGroup struct {
+	Certificate certificateutil.CertificateInfosModel
+	Profiles    map[string]profileutil.ProfileModel
```
Comment (godrei) [inline utils/code_sign_mapping.go]: Profiles -> BundleIDProfileMap

File: utils/code_sign_mapping.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+	"github.com/bitrise-tools/go-xcode/exportoptions"
+	glob "github.com/ryanuber/go-glob"
+
+	"github.com/bitrise-io/steps-certificate-and-profile-installer/certificateutil"
+	"github.com/bitrise-io/steps-certificate-and-profile-installer/profileutil"
+)
+
+// ProfileGroup ...
+type ProfileGroup struct {
```
Comment (godrei) [inline utils/code_sign_mapping.go]: CodeSignGroupItem

File: utils/code_sign_mapping.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	Certificate certificateutil.CertificateInfosModel
+	Profiles    map[string]profileutil.ProfileModel
+}
+
+// ResolveCodeSignMapping ...
+func ResolveCodeSignMapping(bundleIDs []string, exportMethod exportoptions.Method, profiles []profileutil.ProfileModel, certificates []certificateutil.CertificateInfosModel) []ProfileGroup {
```
Comment (godrei) [inline utils/code_sign_mapping.go]: ResolveCodeSignGroupItems

File: utils/code_sign_mapping.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	profileGroups := groupInstalledProfilesByInstalledEmbeddedCertificateSubjects(profiles, certificates)
+	return findProfilesByBundleIDsAndExportMethod(profileGroups, bundleIDs, exportMethod)
+}
+
+func isCertificateInstalled(installedCertificates []certificateutil.CertificateInfosModel, certificate certificateutil.CertificateInfosModel) bool {
+	isCertInstalled := false
```
Comment (godrei) [inline utils/code_sign_mapping.go]: isCertInstalled -> installed

File: utils/code_sign_mapping.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	for _, installedCert := range installedCertificates {
```
Comment (godrei) [inline utils/code_sign_mapping.go]: installedCert -> installedCertificate

File: utils/code_sign_mapping.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+		if certificate.RawSubject == installedCert.RawSubject && certificate.RawEndDate == installedCert.RawEndDate {
+			isCertInstalled = true
+			break
+		}
+	}
+	return isCertInstalled
+}
+
+func groupInstalledProfilesByInstalledEmbeddedCertificateSubjects(profiles []profileutil.ProfileModel, certificates []certificateutil.CertificateInfosModel) map[string][]profileutil.ProfileModel {
```
Comment (godrei) [inline utils/code_sign_mapping.go]: createCertificateProfilesMapping

File: utils/code_sign_mapping.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	groupedProfiles := map[string][]profileutil.ProfileModel{}
```
Comment (godrei) [inline utils/code_sign_mapping.go]: groupedProfiles -> createCertificateProfilesMap

File: utils/code_sign_mapping.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+	for _, profile := range profiles {
+		for _, embeddedCert := range profile.DeveloperCertificates {
```
Comment (godrei) [inline utils/code_sign_mapping.go]: embeddedCert -> Certificate

File: utils/code_sign_mapping.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+			if _, ok := groupedProfiles[embeddedCert.RawSubject]; !ok {
+				groupedProfiles[embeddedCert.RawSubject] = []profileutil.ProfileModel{}
+			}
+			groupedProfiles[embeddedCert.RawSubject] = append(groupedProfiles[embeddedCert.RawSubject], profile)
+		}
+	}
+
+	return groupedProfiles
+}
+
```
Comment (godrei) [inline utils/code_sign_mapping.go]: please create a separate method, which filters the certificateProfilesMap with exportMethod

`func filterCertificateProfilesMap(certificateProfilesMap, exportMethod) certificateProfilesMap `

File: utils/code_sign_mapping.go [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+func findProfilesByBundleIDsAndExportMethod(profileGroups map[string][]profileutil.ProfileModel, bundleIDs []string, exportMethod exportoptions.Method) []ProfileGroup {
```
Comment (godrei) [inline utils/code_sign_mapping.go]: do the exportMethod filtering in filterCertificateProfilesMap

findProfilesByBundleIDsAndExportMethod ->
`func mapCertificateProfilesMapWithBundleIDS(certificateProfilesMap, bundleIDs) [] CodeSignGroupItem `

File: utils/code_sign_mapping.go [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+	installed := false
+	for _, installedCertificate := range installedCertificates {
+		if certificate.Serial == installedCertificate.Serial {
+			installed = true
+			break
+		}
+	}
+
+	if installed {
+		log.Printf("certificate: %s installed", certificate.CommonName)
```
Comment (trapacska) [inline utils/code_sign_mapping.go]: Please remove debug logs, or make them better.

File: utils/code_sign_mapping.go [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+			if _, ok := createCertificateProfilesMap[embeddedCert.Serial]; !ok {
+				createCertificateProfilesMap[embeddedCert.Serial] = []profileutil.ProfileInfoModel{}
+			}
+			createCertificateProfilesMap[embeddedCert.Serial] = append(createCertificateProfilesMap[embeddedCert.Serial], profile)
+		}
+	}
+
+	for subject, profiles := range createCertificateProfilesMap {
+		log.Printf("certificate: %s profiles:", subject)
```
Comment (trapacska) [inline utils/code_sign_mapping.go]: Please remove debug logs, or make them better.

File: utils/code_sign_mapping.go [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+		for _, profile := range profiles {
+			log.Printf("- %s", profile.Name)
```
Comment (trapacska) [inline utils/code_sign_mapping.go]: Please remove debug logs, or make them better.

File: utils/code_sign_mapping.go [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+func createCodeSignGroups(profileGroups map[string][]profileutil.ProfileInfoModel, bundleIDs []string, exportMethod exportoptions.Method) []CodeSignGroupItem {
+	filteredCodeSignGroupItems := []CodeSignGroupItem{}
+	for groupItemCertificateSerial, profiles := range profileGroups {
+		sort.Sort(ByBundleIDLength(profiles))
+
+		bundleIDProfileMap := map[string]profileutil.ProfileInfoModel{}
+		for _, bundleID := range bundleIDs {
+			for _, profile := range profiles {
+				if profile.ExportType != exportMethod {
+					log.Printf("profile: %s is not for export method: %s", profile.Name, exportMethod)
```
Comment (trapacska) [inline utils/code_sign_mapping.go]: Please remove debug logs, or make them better.

File: utils/code_sign_mapping.go [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+					continue
+				}
+
+				if !glob.Glob(profile.BundleIdentifier, bundleID) {
+					log.Printf("profile: %s is not for bundle id: %s", profile.Name, profile.BundleIdentifier)
```
Comment (trapacska) [inline utils/code_sign_mapping.go]: Please remove debug logs, or make them better.

File: utils/code_sign_mapping.go [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+					continue
+				}
+
+				log.Printf("profile: %s MATCHES for: %s", profile.Name, bundleID)
```
Comment (trapacska) [inline utils/code_sign_mapping.go]: Please remove debug logs, or make them better.

File: utils/code_sign_mapping.go [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+
+				bundleIDProfileMap[bundleID] = profile
+				break
+			}
+		}
+
+		log.Printf("len(bundleIDProfileMap): %d <-> len(bundleIDs): %d", len(bundleIDProfileMap), len(bundleIDs))
```
Comment (trapacska) [inline utils/code_sign_mapping.go]: Please remove debug logs, or make them better.

File: utils/code_sign_mapping_test.go [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/bitrise-io/steps-certificate-and-profile-installer/profileutil"
+	"github.com/bitrise-tools/go-xcode/exportoptions"
+	"github.com/stretchr/testify/require"
+)
+
+func TestIsCertificateInstalled(t *testing.T) {
+	t.Log("certificate installed")
+	{
+		certificate := certificateutil.CertificateInfoModel{
+			Serial: "subject= /UID=23442233441/CN=iPhone Developer: INSTALLED (679345FD33)/OU=671115FD33/O=My Company/C=US",
```
Comment (trapacska) [inline utils/code_sign_mapping_test.go]: Please change ex-subjects to dummy serials

**Pull request merged**